create fnc v1.0

darshan3v opened the pull request:
The Core Basic Create function has been created , additional features or bug fixes can be added later on .

darshan3v committed:
create fnc v1.0

artemismars commented on src/crud.rs:
```diff
@@ -0,0 +1,56 @@
+
+impl CPath {
+    pub fn new(path: &str) -> Result<Self, io::ErrorKind> {
+        let n = path.len();
```
`let max_index = path.len() - 1`

artemismars replied:
```suggestion
        let n = path.len();
```

artemismars commented on src/crud.rs:
```diff
@@ -0,0 +1,56 @@
+    pub fn new(path: &str) -> Result<Self, io::ErrorKind> {
+        let n = path.len();
+        let i = path.find(".").unwrap_or_default();
+        if i == 0 || i == n - 1 || i != path.rfind(".").unwrap() {
```
I think `if` condition could be more intuitive.

artemismars replied:
```suggestion
        let max_index = path.len() - 1;
        let first = path.find(".").unwrap_or_default();
        if first == 0 || first == max_index || first != path.rfind(".").unwrap() {
```

artemismars replied:
.

artemismars commented on src/crud.rs:
```diff
@@ -0,0 +1,56 @@
+
+pub fn create_file(file_name: &str) -> Result<File, io::ErrorKind> {
+    let path = CPath::new(file_name)?;
+    match File::create(path.path) {
```
`file_name` could be `path` since the input it's not only for filename but can be the entire file path.

artemismars replied:
```suggestion
pub fn create_file(path: &str) -> Result<File, io::ErrorKind> {
    let path = CPath::new(path)?.path;
    match File::create(path) {
```
now the struct `Cpath` deals with `path`. I though this makes more sense.
also `?` operator gives us the inner value of `Ok(v)` So we can access `cpath` in `Cpath` struct right away.

artemismars commented on src/crud.rs:
```diff
@@ -0,0 +1,56 @@
+use std::{fs::File, io};
+
+struct CPath {
+    path: String,
```
I think the key `path` should be `cpath`. You can check the reason on line 23.

artemismars replied:
```suggestion
    cpath: String,
```

artemismars commented on src/crud.rs:
```diff
@@ -0,0 +1,56 @@
+    #[should_panic]
+    fn test_create_file_f3() {
+        create_file("hello").unwrap();
+    }
```
instead of creating each failure functions we can put all of them into one failure function by using `unwrap_err` function.

artemismars replied:
```suggestion
    fn test_create_file_fail() {
        create_file("hello.t.xt").unwrap_err();
        create_file("hello.").unwrap_err();
        create_file("hello").unwrap_err();
    }
```
If `create_file` function returns Err then `unwrap_err` function returns error message.
If `create_file` function returns Ok(v) then `unwrap_err` panics at `v`.
So `unwrap_err` will panics when `create_file` works normally.

artemismars commented:
hi

darshan3v committed:
Increasing Readability and making code concise

Made if condition look more intuitive ,
made create file function more concise,
merged all the fail test in one function

Co-authored-by: artemismars <75290655+artemismars@users.noreply.github.com>

artemismars commented:
re-open
